txnsync/txngroups.go

```diff
@@ -60,6 +454,953 @@ func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGro
+	}
+	if txn.Txn.Note != nil && len(txn.Txn.Note) > 0 {
+		stub.BitmaskNote.SetBit(i)
+		stub.Note = append(stub.Note, txn.Txn.Note)
+	}
+	if txn.Txn.GenesisID != "" {
+		stub.BitmaskGenesisID.SetBit(i)
+		stub.GenesisID = txn.Txn.GenesisID
+	}
+	if !txn.Txn.GenesisHash.MsgIsZero() {
+		stub.BitmaskGenesisHash.SetBit(i)
+		stub.GenesisHash = txn.Txn.GenesisHash
+	}
+	if !txn.Txn.Group.MsgIsZero() {
+		stub.BitmaskGroup.SetBit(i)
+	}
+	if txn.Txn.Lease != ([32]byte{}) {
+		stub.BitmaskLease.SetBit(i)
+		stub.Lease = append(stub.Lease, txn.Txn.Lease[:]...)
+	}
+	if !txn.Txn.RekeyTo.MsgIsZero() {
+		stub.BitmaskRekeyTo.SetBit(i)
+		stub.RekeyTo = append(stub.RekeyTo, txn.Txn.RekeyTo[:]...)
+	}
+}
+
+func finishDeconstructTxnHeader(stub *txGroupsEncodingStub) {
+	stub.BitmaskSender.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskFee.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskFirstValid.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskLastValid.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskNote.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskGenesisID.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskGenesisHash.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskGroup.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskLease.trimBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskRekeyTo.trimBitmask(int(stub.TotalTransactionsCount))
+}
+
+func setupDeconstructKeyregTxnFields(stub *txGroupsEncodingStub) {
+	bitmaskLen := bytesNeededBitmask(int(stub.TotalTransactionsCount))
+	stub.BitmaskVotePK = make(bitmask, bitmaskLen)
+	stub.BitmaskSelectionPK = make(bitmask, bitmaskLen)
+	stub.BitmaskVoteFirst = make(bitmask, bitmaskLen)
+	stub.BitmaskVoteLast = make(bitmask, bitmaskLen)
+	stub.BitmaskVoteKeyDilution = make(bitmask, bitmaskLen)
+	stub.BitmaskNonparticipation = make(bitmask, bitmaskLen)
+}
+
+func deconstructKeyregTxnFields(stub *txGroupsEncodingStub, i int, txn transactions.SignedTxn) {
+	if txn.Txn.Type == protocol.KeyRegistrationTx {
+		if !txn.Txn.VotePK.MsgIsZero() {
+			stub.BitmaskVotePK.SetBit(i)
+			stub.VotePK = append(stub.VotePK, txn.Txn.VotePK[:]...)
+		}
+		if !txn.Txn.SelectionPK.MsgIsZero() {
+			stub.BitmaskSelectionPK.SetBit(i)
+			stub.SelectionPK = append(stub.SelectionPK, txn.Txn.SelectionPK[:]...)
+		}
+		if !txn.Txn.VoteFirst.MsgIsZero() {
+			stub.BitmaskVoteFirst.SetBit(i)
+			stub.VoteFirst = append(stub.VoteFirst, txn.Txn.VoteFirst)
+		}
+		if !txn.Txn.VoteLast.MsgIsZero() {
+			stub.BitmaskVoteLast.SetBit(i)
+			stub.VoteLast = append(stub.VoteLast, txn.Txn.VoteLast)
+		}
+		if txn.Txn.VoteKeyDilution > 0 {
+			stub.BitmaskVoteKeyDilution.SetBit(i)
+			stub.VoteKeyDilution = append(stub.VoteKeyDilution, txn.Txn.VoteKeyDilution)
+		}
+		if txn.Txn.Nonparticipation {
+			stub.BitmaskNonparticipation.SetBit(i)
+			stub.Nonparticipation = append(stub.Nonparticipation, txn.Txn.Nonparticipation) // can probably get rid of this
```

Comment:
given that it's a boolean the above bitset should be enough.

Reply:
yes, I put a comment reminding myself to get rid of that